### ISSUE TITLE
Serious memory leak: Ammo btRigidBodies are (almost) never destroyed

### ISSUE THREAD

**Issue by ThatStevenGuy** [playcanvas/engine#5801]
### Description
Hey everyone,

We have very large worlds in one of our games with roughly 3-5k rigid bodies per scene (only a handful are active at a time). We noticed after switching scenes a couple of times, Ammo would eventually crash with an `abort(OOM)`-error indicating it exhausted its 64 Mb heap. After some digging, it turns out almost none of Ammo's underlying `btRigidBody`'s in our scenes are actually destroyed.

### Cause
The issue stems from `RigidBodyComponentSystem`. This class listens for the removal of rigid body components so it can destroy the underlying `btRigidBody`. It does this by binding its `onRemove`-method to the `remove`-event of `ComponentSystem` during initialization. This `remove`-event however requires a signature of `(entity: Entity, componentData: any)`, but the `onRemove`-method has a signature of `(entity: Entity, component: RigidBodyComponent)` (note the type of the second parameter). 

As a result, whenever a rigid body component is destroyed and the `remove`-event is fired, `RigidBodyComponentSystem.onRemove` will receive a non-component argument for its `component`-parameter. This causes the destruction of the underlying body to always fail, since the method can't find the `body`-field on what it thinks is a rigid body component, but is actually an argument of type `RigidBodyComponentData`.

Small side note: the only time `RigidBodyComponentSystem.onRemove` works as expected is through `RigidBodyComponent.createBody()`, which invokes `onRemove` to destroy any previous `btRigidBody` instance.

### Steps to Reproduce
1. In the Editor, create a new scene;
2. Under the `Root` entity, find the default `Box` entity and attach a collision component and a rigid body component to it (default values are fine);
3. Press play. Make sure the URL includes `debug=true`;
4. Open your browser's developer tools and find the file `rigid-body/system.js` which contains `RigidBodyComponentSystem`. Place a breakpoint on the first line of the `onRemove`-method;
5. Destroy the box by invoking `pc.app.root.children[0].children[2].destroy();` in the JavaScript console;
6. The breakpoint you've set will be hit. Note that the `component`-argument is in fact of the type `RigidBodyComponentData` which only contains an `enabled`-field, not a `body`-field. The check for the existence of `component.body` will always fail, and the body will remain in Ammo's memory;

### Quick-Fix
A quick way to resolve the issue is to make a new script named `fix.js` and insert the following:

```
let Fix = pc.createScript('fix');

Fix.prototype.initialize = function() {
    const rigidBodyComponentSystem = this.app.systems.rigidbody;
    rigidBodyComponentSystem.off('beforeremove', rigidBodyComponentSystem.onBeforeRemove, rigidBodyComponentSystem);
    rigidBodyComponentSystem.on('beforeremove', (_entity, component) => {
        component.data.body = component.body; // Ensures component.body access in onRemove() functions as expected

        if(component.enabled) {
            component.enabled = false;
        }
    });

    console.log("Patch applied.");
};
```
Add this script to any entity in the scene (`Root` is fine) and run the scene again. Follow the same steps as above, and you'll find that this time the `onRemove`-method successfully finds `component.body` and destroys the body accordingly:

![Fixed](https://github.com/playcanvas/engine/assets/65724470/25ef587f-589e-46a3-ac79-1d5df14976a0)

This script rebinds `RigidBodyComponentSystem.onBeforeRemove` so that it copies over the  `body`-field from the `RigidBodyComponent` to the component's `RigidBodyComponentData`. This ensures the `onRemove`-method works as expected.

Thanks for looking into the issue. This one took a while to find.

**Comment by LeXXik** [comment]
Duplicate to #4145

**Comment by marklundin** [comment]
Thanks for the detailed digging @ThatStevenGuy. This has now been fixed in #5803

**Comment by ThatStevenGuy** [comment]
> Thanks for the detailed digging @ThatStevenGuy. This has now been fixed in #5803

Awesome! Thanks a bunch for addressing the issue :muscle:
